lib/apimlStorage.ts

```diff
@@ -0,0 +1,382 @@
+  });
+}
+
+export function isConfigured(): boolean {
+  return apimlClient != null;
+}
+
+export interface ApimlStorageSettings {
+  host: string;
+  port: number;
+  tlsOptions: https.AgentOptions;
+}
+
+
+export function makeStorageForPlugin(pluginId: string): ApimlStorage {
+  return new ApimlStorage(pluginId);
+}
+
+const CACHING_SERVICE_URI = '/cachingservice/api/v1/cache';
+
+interface ApimlRequest {
+  method: 'GET' | 'POST' | 'PUT' | 'DELETE',
+  path: string;
+  headers?: { [key: string]: string };
+  body?: any;
+}
+
+interface KeyValuePair {
+  key: string;
+  value: any;
+}
+interface ApimlResponse {
+  statusCode: number;
+  json?: any;
+  headers: http.IncomingHttpHeaders;
+}
+
+export type ApimlStorageErrorCode =
+  'APIML_STORAGE_CONNECTION_ERROR' |
+  'APIML_STORAGE_KEY_NOT_FOUND' |
+  'APIML_STORAGE_JSON_ERROR' |
+  'APIML_STORAGE_UNKNOWN_CLIENT_CERT' |
+  'APIML_STORAGE_UNAUTHORIZED' |
+  'APIML_STORAGE_NO_AUTH_PROVIDED' |
+  'APIML_STORAGE_INVALID_PAYLOAD' |
+  'APIML_STORAGE_UNKNOWN_ERROR' |
+  'APIML_STORAGE_RESPONSE_ERROR' |
+  'APIML_STORAGE_NOT_CONFIGURED';
+
+export class ApimlStorageError extends Error {
+  constructor(
+    public readonly code: ApimlStorageErrorCode,
+    private cause?: Error,
+    private apimlResponse?: ApimlResponse) {
+    super(code);
+    Object.setPrototypeOf(this, ApimlStorageError.prototype);
+  }
+
+  toString(): string {
+    const apimlMessages = this.getApimlMessages();
+    const errorMessage = this.cause ? this.cause.message : undefined;
+    let resultMessage = this.code;
+    if (errorMessage) {
+      resultMessage += ' ' + errorMessage;
+    }
+    if (apimlMessages) {
+      resultMessage += ' ' + apimlMessages;
+    }
+    return resultMessage;
+  }
+
+  private getApimlMessages(): string | undefined {
+    if (typeof this.apimlResponse !== 'object') {
+      return undefined;
+    }
+    if (!isApimlErrorMessages(this.apimlResponse.json)) {
+      return undefined;
+    }
+    return this.apimlResponse.json.messages.map(msg => getApimlErrorMessageString(msg)).join(';');
+  }
+
+}
+
+export function isApimlStorageError(e: Error): e is ApimlStorageError {
+  return e instanceof ApimlStorageError;
+}
+
+export function isApimlStorageKeyNotFoundError(e: Error) {
+  return isApimlStorageError(e) && e.code === 'APIML_STORAGE_KEY_NOT_FOUND';
+}
+
+export interface ApimlErrorMessage {
+  messageType: 'ERROR' | 'INFO' | 'WARN';
+  messageNumber: string;
+  messageContent: string;
+  messageKey: string;
+}
+export interface ApimlErrorMessages {
+  messages: ApimlErrorMessage[];
+}
+
+function isApimlErrorMessages(obj: any): obj is ApimlErrorMessages {
+  if (typeof obj !== 'object') {
+    return false;
+  }
+  if (!Array.isArray(obj.messages)) {
+    return false;
+  }
+  return true;
+}
+
+function getApimlErrorMessageString(message: ApimlErrorMessage): string {
+  return `${message.messageType}(${message.messageKey}) ${message.messageNumber} ${message.messageContent}`;
```

Comment:
When the message is printed in the zlux logger, there will be two message types and it could be confusing.
If someone did a log.info on a message type error, it would say
timestamp (zwed:pid) user INFO - ERROR
The message key being inside () is also not the existing logger format.
Maybe this file should just call the zlux logger directly instead of returning toString() results.

Reply:
I removed the `messageType` from the message